### PR TITLE
Print meaningful error when too much shared memory is allocated on CUDA

### PR DESCRIPTION
**Pull request by KoolJBlack**
Queries  the driver for the max amount of shared memory allowed then errors if too much is requested. 

Addresses #12146

### CODE REVIEW

File: runtime/src/iree/hal/drivers/cuda/native_executable.c [hunk 1]
```diff
@@ -120,12 +120,26 @@ iree_status_t iree_hal_cuda_native_executable_create(
         break;
       }
 
+      int32_t max_shared_mem;
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/cuda/native_executable.c]: always zero initialize (`= 0;`) - locals are undefined by default

File: runtime/src/iree/hal/drivers/cuda/native_executable.c [hunk 2]
```diff
@@ -120,12 +120,26 @@ iree_status_t iree_hal_cuda_native_executable_create(
...
           context->syms,
-          cuFuncSetAttribute(function,
-                             CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES,
-                             shared_memory_sizes[i]),
-          "cuFuncSetAttribute");
+          cuDeviceGetAttribute(&max_shared_mem,
+                               CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_BLOCK,
+                               context->cu_device),
+          "cuDeviceGetAttribute");
+      if (shared_memory_sizes[i] > max_shared_mem) {
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/cuda/native_executable.c:132]: this leaks/ignores status - you need to do an `if (!iree_status_is_ok(status)) break;` after each `status =`

**Review by benvanik — CHANGES_REQUESTED** [review]

**Comment by powderluv** [comment]
Do we need the same for SPIR-V like in https://github.com/openxla/iree/issues/12523 ?

**Review by benvanik — APPROVED** [review]

**Comment by benvanik** [comment]
We should be able to get some more detailed diagnostics on both CUDA/SPIR-V - I'll look at plumbing that through (we can ask the driver compilers for full logs and attach those to our errors). This is still nice as a concise error message but the general one will work to avoid needing a bunch of these kind of things.

**Pull request merged**